File: backend/schemas/benchmarking.py
import uuid
from datetime import datetime

from pydantic import BaseModel, Field


class BenchmarkAccountCreateRequest(BaseModel):
    client_id: uuid.UUID
    platform: str
    handle: str
    purpose: str = "all"
    source_type: str = "manual"
    memo: str | None = None
    metadata_json: dict | None = None


class BenchmarkAccountUpdateRequest(BaseModel):
    handle: str | None = None
    purpose: str | None = None
    source_type: str | None = None
    memo: str | None = None
    metadata_json: dict | None = None
    is_active: bool | None = None


class BenchmarkAccountResponse(BaseModel):
    id: uuid.UUID
    client_id: uuid.UUID
    platform: str
    handle: str
    source_type: str = "manual"
    purpose: str = "all"
    memo: str | None = None
    auto_discovered: bool = False
    is_active: bool = True
    metadata_json: dict | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None

    model_config = {"from_attributes": True}


class BenchmarkPostResponse(BaseModel):
    id: uuid.UUID
    benchmark_account_id: uuid.UUID
    client_id: uuid.UUID
    platform: str
    external_post_id: str | None = None
    post_url: str | None = None
    content_text: str | None = None
    hook_text: str | None = None
    cta_text: str | None = None
    format_type: str | None = None
    view_count: int = 0
    like_count: int = 0
    comment_count: int = 0
    share_count: int = 0
    save_count: int = 0
    engagement_rate: float = 0.0
    benchmark_score: float = 0.0
    published_at: datetime | None = None
    raw_payload: dict | None = None
    source_scope: str = "client_direct"
    source_scope_label: str = "직접 클라이언트"
    is_direct_client_post: bool = True

    model_config = {"from_attributes": True}


class RefreshAccountResponse(BaseModel):
    status: str
    status_label: str | None = None
    message: str
    inserted: int = 0
    profile_id: uuid.UUID | None = None
    profile_generated: bool = False
    live_supported: bool = False
    platform: str | None = None
    used_placeholder: bool = False
    data_source: str | None = None
    data_source_label: str | None = None
    view_metric_type: str | None = None
    view_metric_label: str | None = None
    source_channel_connected: bool = False
    source_channel_platform: str | None = None
    source_channel_account_name: str | None = None
    source_channel_missing_reason: str | None = None
    source_channel_has_token: bool = False
    source_channel_connection_count: int = 0
    source_channel_duplicate_count: int = 0
    source_channel_duplicate_warning: bool = False
    refreshed_at: datetime | None = None


class BenchmarkAccountDiagnosticResponse(BaseModel):
    account_id: uuid.UUID
    client_id: uuid.UUID
    platform: str
    handle: str
    is_active: bool = True
    support_level: str
    support_label: str
    status: str
    status_label: str | None = None
    message: str
    live_supported: bool = False
    source_channel_connected: bool = False
    source_channel_platform: str | None = None
    source_channel_account_name: str | None = None
    source_channel_missing_reason: str | None = None
    source_channel_has_token: bool = False
    source_channel_connection_count: int = 0
    source_channel_duplicate_count: int = 0
    source_channel_duplicate_warning: bool = False
    live_post_count: int = 0
    placeholder_post_count: int = 0
    actual_metric_count: int = 0
    proxy_metric_count: int = 0
    total_post_count: int = 0
    data_source: str | None = None
    data_source_label: str | None = None
    view_metric_type: str | None = None
    view_metric_label: str | None = None
    used_placeholder: bool = False
    last_refresh_status: str | None = None
    last_refresh_status_label: str | None = None
    last_refresh_message: str | None = None
    last_refresh_inserted: int = 0
    last_refresh_profile_id: uuid.UUID | None = None
    last_refresh_profile_generated: bool = False
    last_refresh_used_placeholder: bool = False
    last_refresh_data_source: str | None = None
    last_refresh_data_source_label: str | None = None
    last_refresh_view_metric_type: str | None = None
    last_refresh_view_metric_label: str | None = None
    last_refresh_at: datetime | None = None


class ActionLanguageProfileResponse(BaseModel):
    id: uuid.UUID | None = None
    client_id: uuid.UUID
    platform: str
    source_scope: str
    top_hooks_json: list | None = None
    top_ctas_json: list | None = None
    tone_patterns_json: dict | None = None
    format_patterns_json: dict | None = None
    recommended_prompt_rules: str | None = None
    profile_version: int = 1
    updated_at: datetime | None = None
    source_client_id: uuid.UUID | None = None
    industry_category: str | None = None
    sample_count: int = 0

    model_config = {"from_attributes": True}


class TopPostsQuery(BaseModel):
    client_id: uuid.UUID
    platform: str
    top_k: int = Field(default=10, ge=1, le=50)

File: frontend/src/services/benchmarking.ts
import api from "./api"

export interface BenchmarkAccountItem {
  id: string
  client_id: string
  platform: string
  handle: string
  source_type: string
  purpose: string
  memo?: string | null
  auto_discovered: boolean
  is_active: boolean
  metadata_json?: Record<string, unknown> | null
  updated_at?: string | null
}

export interface BenchmarkPostItem {
  id: string
  benchmark_account_id: string
  platform: string
  external_post_id?: string | null
  post_url?: string | null
  content_text?: string | null
  hook_text?: string | null
  cta_text?: string | null
  format_type?: string | null
  view_count: number
  like_count: number
  comment_count: number
  share_count: number
  save_count: number
  engagement_rate: number
  benchmark_score: number
  published_at?: string | null
  raw_payload?: Record<string, unknown> | null
}

export interface ActionLanguageProfileItem {
  id?: string | null
  client_id: string
  platform: string
  source_scope: string
  top_hooks_json?: Array<{ pattern: string; count: number }> | null
  top_ctas_json?: Array<{ pattern: string; count: number }> | null
  tone_patterns_json?: Record<string, unknown> | null
  format_patterns_json?: Record<string, unknown> | null
  recommended_prompt_rules?: string | null
  profile_version: number
  updated_at?: string | null
  source_client_id?: string | null
  industry_category?: string | null
  sample_count?: number
}

export interface RefreshAccountResult {
  status: string
  status_label?: string | null
  message: string
  inserted: number
  profile_id?: string | null
  profile_generated?: boolean
  live_supported?: boolean
  platform?: string
  used_placeholder?: boolean
  data_source?: string | null
  data_source_label?: string | null
  view_metric_type?: string | null
  view_metric_label?: string | null
  source_channel_connected?: boolean
  source_channel_platform?: string | null
  source_channel_account_name?: string | null
  source_channel_missing_reason?: string | null
  source_channel_has_token?: boolean
  refreshed_at?: string | null
}

export interface BenchmarkAccountDiagnosticItem {
  account_id: string
  client_id: string
  platform: string
  handle: string
  is_active: boolean
  support_level: string
  support_label: string
  status: string
  status_label?: string | null
  message: string
  live_supported: boolean
  source_channel_connected: boolean
  source_channel_platform?: string | null
  source_channel_account_name?: string | null
  source_channel_missing_reason?: string | null
  source_channel_has_token: boolean
  live_post_count: number
  placeholder_post_count: number
  actual_metric_count: number
  proxy_metric_count: number
  total_post_count: number
  data_source?: string | null
  data_source_label?: string | null
  view_metric_type?: string | null
  view_metric_label?: string | null
  used_placeholder: boolean
  last_refresh_status?: string | null
  last_refresh_status_label?: string | null
  last_refresh_message?: string | null
  last_refresh_inserted?: number
  last_refresh_profile_id?: string | null
  last_refresh_profile_generated?: boolean
  last_refresh_used_placeholder?: boolean
  last_refresh_data_source?: string | null
  last_refresh_data_source_label?: string | null
  last_refresh_view_metric_type?: string | null
  last_refresh_view_metric_label?: string | null
  last_refresh_at?: string | null
}

export const benchmarkingService = {
  async listAccounts(clientId?: string) {
    const res = await api.get("/api/v1/benchmarking/accounts", { params: clientId ? { client_id: clientId } : {} })
    return Array.isArray(res.data) ? (res.data as BenchmarkAccountItem[]) : []
  },

  async createAccount(payload: { client_id: string; platform: string; handle: string; purpose?: string; source_type?: string; memo?: string; metadata_json?: Record<string, unknown> }) {
    const res = await api.post("/api/v1/benchmarking/accounts", payload)
    return res.data as BenchmarkAccountItem
  },

  async updateAccount(id: string, payload: Partial<BenchmarkAccountItem>) {
    const res = await api.patch(`/api/v1/benchmarking/accounts/${id}`, payload)
    return res.data as BenchmarkAccountItem
  },

  async refreshAccount(id: string, topK = 10, windowDays = 30) {
    const res = await api.post(`/api/v1/benchmarking/accounts/${id}/refresh`, null, {
      params: { top_k: topK, window_days: windowDays },
    })
    return res.data as RefreshAccountResult
  },

  async listAccountDiagnostics(clientId: string, platform?: string) {
    const res = await api.get("/api/v1/benchmarking/account-diagnostics", {
      params: {
        client_id: clientId,
        ...(platform ? { platform } : {}),
      },
    })
    return Array.isArray(res.data) ? (res.data as BenchmarkAccountDiagnosticItem[]) : []
  },

  async getTopPosts(clientId: string, platform: string, topK = 10) {
    const res = await api.get("/api/v1/benchmarking/top-posts", {
      params: { client_id: clientId, platform, top_k: topK },
    })
    return Array.isArray(res.data) ? (res.data as BenchmarkPostItem[]) : []
  },

  async getActionProfile(clientId: string, platform: string) {
    const res = await api.get("/api/v1/benchmarking/action-language-profile", {
      params: { client_id: clientId, platform },
    })
    return (res.data || null) as ActionLanguageProfileItem | null
  },
}
